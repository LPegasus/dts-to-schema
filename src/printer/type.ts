import type * as ts from "typescript";

export interface PrintAdaptor {
  emitSchema(printSourceData: PrintSourceData): Promise<void>;
  emitEntry(
    outputFileName: string,
    printSourceDataCollection: PrintSourceData[]
  ): Promise<void>;
}

export enum DtsModuleType {
  Legacy = "legacy",
  Module = "module",
}

export type PrintSourceData = PrintSourceEnumData | PrintSourceInterfaceData;

export interface PrintSourceDataBase {
  /** 输出目录，输出目录通过 namespace + outDir 决定 */
  outputDir: string;
  moduleType: DtsModuleType;
  fullyQualifiedName: string;
  /** TS 定义所在文件名，生成的 schema 文件需要类型定义 */
  relativeDefinitionFileName?: string;
  /** interface 所有成员，string 表示是原始类型，PrintTypeReference 表示用 ref */
  kind: ts.SyntaxKind;
}

export interface PrintSourceEnumData extends PrintSourceDataBase {
  kind: ts.SyntaxKind.EnumDeclaration;
  members: Array<{ name: string; value: string | number }>;
}

export interface PrintSourceInterfaceData extends PrintSourceDataBase {
  kind: ts.SyntaxKind.InterfaceDeclaration;
  props: Record<string, PrintMemberType>;
}

export interface BasePrintMemberType {
  v: string | PrintTypeReference | "any";
  optional: boolean;
  nullable: boolean;
}

export interface PrintMemberTypeOfArray extends BasePrintMemberType {
  isArray: true;
  latitude: number;
}

export interface PrintMemberTypeOfObject extends BasePrintMemberType {
  isArray: false;
}

export type PrintMemberType = PrintMemberTypeOfArray | PrintMemberTypeOfObject;

export interface PrintTypeReference {
  ref: string;
}

export enum PrinterEmitSchemaType {
  JTD = "jtd",
  JSON = "json",
}

export interface PrinterConfiguration {
  /** output directory */
  outDir?: string;
  emitSchemaType?: PrinterEmitSchemaType;
  prettierConfig?: import("prettier").Options;
}

export interface PrintHost {
  writeFileAsync(fileName: string, content: string): Promise<void>;
  getOutputAbsoluteFileName(printSourceData: PrintSourceData): string;
  normalizeVariableNameFromQualifiedName(fullyQualifiedName: string): string;
}
