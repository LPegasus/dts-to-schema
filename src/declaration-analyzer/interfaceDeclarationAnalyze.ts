import * as ts from "typescript";
import {
  DtsModuleType,
  PrintMemberType,
  PrintMemberTypeOfArray,
  PrintSourceData,
  PrintSourceInterfaceData,
} from "../printer/type";
import * as path from "path";
import { baseAnalyze } from "./baseAnalyze";

export function interfaceDeclarationAnalyze(
  typeChecker: ts.TypeChecker,
  node: ts.InterfaceDeclaration,
  outputDir: string,
  accumulator: Map<string, PrintSourceData>
): PrintSourceInterfaceData {
  const { symbol, fullyQualifiedName, lastDeclaration } = baseAnalyze(
    node,
    typeChecker
  );
  if (accumulator.has(fullyQualifiedName)) {
    return accumulator.get(fullyQualifiedName)! as any;
  }

  const splittedQualifiedNames = fullyQualifiedName.split(".");
  const emitDir = path.resolve(
    outputDir,
    ...splittedQualifiedNames.slice(0, -1)
  );

  const printSourceData: PrintSourceData = {
    kind: node.kind,
    fullyQualifiedName,
    outputDir: emitDir,
    relativeDefinitionFileName: path.relative(
      outputDir,
      lastDeclaration.getSourceFile().fileName
    ),
    moduleType: DtsModuleType.Legacy,
    props: {},
  };
  symbol.members?.forEach((member) => {
    const declarations = member.getDeclarations();
    // only choose the last declaration;
    const declaration = declarations![declarations!.length - 1];
    if (ts.isMethodSignature(declaration)) {
      // ignore method property
      return;
    }
    if (ts.isPropertySignature(declaration)) {
      if (!declaration.type) {
        // ignore property without type signature
        return;
      }
      if (!ts.isIdentifier(declaration.name)) {
        // ignore index signature key, eg: {[key: string]: Type}
        return;
      }
      const optional = Boolean(declaration.questionToken);
      const data = getPrintSourceDataFromTypeNode(
        declaration.type,
        typeChecker
      );
      if (!data) {
        return;
      }
      printSourceData.props[declaration.name.escapedText!] = {
        ...data,
        nullable: false,
        optional,
      } as PrintMemberType;
    }
    return;
  });
  accumulator.set(fullyQualifiedName, printSourceData);
  return printSourceData;
}

function getPrintSourceDataFromTypeNode(
  node: ts.TypeNode,
  typeChecker: ts.TypeChecker
): Omit<PrintMemberType, "optional" | "nullable"> | null {
  if (ts.isTypeReferenceNode(node)) {
    return getPrintSourceDataFromTypeReferenceNode(node, typeChecker);
  }
  if (ts.isArrayTypeNode(node)) {
    return getPrintSourceDataFromArrayTypeNode(node, typeChecker);
  }
  if (ts.isToken(node)) {
    return {
      v: node.getText(),
      isArray: false,
    };
  }
  return null;
}

function getPrintSourceDataFromTypeReferenceNode(
  node: ts.TypeReferenceNode,
  typeChecker: ts.TypeChecker
): Omit<PrintMemberType, "optional" | "nullable"> | null {
  if ((node.typeName as ts.Identifier).getText() === "Array") {
    const t = node.typeArguments?.[0];
    if (!t) {
      throw new Error(`Array should have type argument.`);
    }
    const inner = getPrintSourceDataFromTypeNode(t, typeChecker);

    if (!inner) {
      return inner;
    }

    if ((inner as PrintMemberTypeOfArray).latitude === undefined) {
      (inner as PrintMemberTypeOfArray).latitude = 0;
    }

    (inner as PrintMemberTypeOfArray).latitude++;
    return {
      ...inner,
      isArray: true,
    };
  } else {
    const nodeType = typeChecker.getTypeFromTypeNode(node);
    const symbol = nodeType.getSymbol();
    if (!symbol) {
      throw new Error(
        `Type reference should have symbol. Check type "${node.typeName.getText()}" whether declared.`
      );
    }
    return {
      v: {
        ref: typeChecker.getFullyQualifiedName(symbol),
      },
      isArray: false,
    };
  }
}

function getPrintSourceDataFromArrayTypeNode(
  node: ts.ArrayTypeNode,
  typeChecker: ts.TypeChecker
): Omit<PrintMemberType, "optional" | "nullable"> | null {
  const t = node.elementType;
  if (!t) {
    throw new Error(`Array should have type argument.`);
  }
  const inner = getPrintSourceDataFromTypeNode(t, typeChecker);

  if (!inner) {
    return inner;
  }

  if ((inner as any).latitude === undefined) {
    (inner as PrintMemberTypeOfArray).latitude = 0;
  }
  (inner as PrintMemberTypeOfArray).latitude++;

  return {
    ...inner,
    isArray: true,
  };
}
