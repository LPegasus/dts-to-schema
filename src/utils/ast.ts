import * as ts from "typescript";

export function printImportStartStatement(
  aliasName: string,
  moduleSpecifier: string
): string {
  const ast = ts.factory.createImportDeclaration(
    undefined,
    undefined,
    ts.factory.createImportClause(
      false,
      undefined,
      ts.factory.createNamespaceImport(ts.factory.createIdentifier(aliasName))
    ),
    ts.factory.createStringLiteral(moduleSpecifier),
    undefined
  );

  const code = ts
    .createPrinter({ newLine: ts.NewLineKind.LineFeed })
    .printNode(
      ts.EmitHint.Expression,
      ast,
      ts.createSourceFile("", "", ts.ScriptTarget.ESNext, false)
    );

  return code;
}

export function filterNodeForSourceFile(
  sourceFile: ts.SourceFile,
  predicate: (node: ts.Node) => boolean
) {
  const results: ts.Node[] = [];
  const scan = (node: ts.Node) => {
    if (predicate(node)) {
      results.push(node);
    }
    node.forEachChild(scan);
  };

  scan(sourceFile);

  return results;
}
