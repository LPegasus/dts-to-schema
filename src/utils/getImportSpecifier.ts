import type { DtsModuleType } from "../printer/type";

export function getImportSpecifier(
  fileName: string,
  dtsModuleType: DtsModuleType
) {}
