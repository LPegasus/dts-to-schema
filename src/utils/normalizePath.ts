export function normalizePath(path: string) {
  let p = path.replace(/\\/g, "/");

  if (process.platform === "win32") {
    p = p.replace(/^([a-zA-Z]:)/, "/$1");
  }

  if (!/^[./]/.test(p)) {
    p = "./" + p;
  }
  return p;
}
