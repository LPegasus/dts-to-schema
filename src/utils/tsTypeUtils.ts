import * as ts from "typescript";

export function getEnumValueType(node: ts.EnumDeclaration): ts.SyntaxKind[] {
  return node.members.map((member) => {
    return member.initializer?.kind ?? ts.SyntaxKind.NumericLiteral;
  });
}

export function isEnumMembersAllNumber(node: ts.EnumDeclaration): boolean {
  return getEnumValueType(node).every(
    (type) => type === ts.SyntaxKind.NumericLiteral
  );
}
