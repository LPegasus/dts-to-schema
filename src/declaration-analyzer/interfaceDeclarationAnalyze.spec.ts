import { interfaceDeclarationAnalyze } from "./interfaceDeclarationAnalyze";
import * as ts from "typescript";
import { filterNodeForSourceFile } from "../utils/ast";
import type { PrintSourceData } from "../printer/type";

const compilerOptions = {
  module: ts.ModuleKind.ESNext,
  moduleResolution: ts.ModuleResolutionKind.NodeJs,
  target: ts.ScriptTarget.ESNext,
  types: [],
};
const compilerHost = ts.createCompilerHost(compilerOptions, true);

const testSnaps: Record<string, string> = {
  "base.d.ts": `interface Base {
    foo: string;
    optionalField?: number;
  };`,
  "array.d.ts": `interface TestArray {
    arr1: string[];
    arr2: string[][];
    arr3: Array<number>;
    arr4: Array<Array<number>>;
    arr5: Array<number[]>;
  };`,
};

compilerHost.fileExists = () => true;
compilerHost.readFile = (fileName: string) => {
  return testSnaps[fileName];
};

describe("interface declaration analyze", () => {
  let prog: ts.Program;
  let typeChecker: ts.TypeChecker;

  beforeEach(() => {
    prog = ts.createProgram({
      rootNames: ["base.d.ts", "array.d.ts"],
      options: {
        module: ts.ModuleKind.ESNext,
        moduleResolution: ts.ModuleResolutionKind.NodeJs,
        target: ts.ScriptTarget.ESNext,
        types: [],
      },
      host: compilerHost,
    });

    typeChecker = prog.getTypeChecker();
  });

  test("base interface", () => {
    const sourceFile = prog.getSourceFile("base.d.ts")!;
    const interfaceNodeList = filterNodeForSourceFile(sourceFile, (n) =>
      ts.isInterfaceDeclaration(n)
    );
    expect(interfaceNodeList.length).toEqual(1);

    const node = interfaceNodeList[0];
    const cache = new Map<string, PrintSourceData>();
    const printSourceData = interfaceDeclarationAnalyze(
      typeChecker,
      node as ts.InterfaceDeclaration,
      "",
      cache
    );

    expect(printSourceData).toMatchObject({
      fullyQualifiedName: "Base",
      kind: ts.SyntaxKind.InterfaceDeclaration,
    });

    expect(printSourceData.props).toMatchSnapshot("base");

    expect(cache.size).toEqual(1);
  });

  test("array.d.ts", () => {
    const sourceFile = prog.getSourceFile("array.d.ts")!;
    const node = filterNodeForSourceFile(sourceFile, (n) => {
      return ts.isInterfaceDeclaration(n);
    })[0] as ts.InterfaceDeclaration;
    const cache = new Map<string, PrintSourceData>();
    const printSourceData = interfaceDeclarationAnalyze(
      typeChecker,
      node,
      "",
      cache
    );

    expect(printSourceData).toMatchObject({
      fullyQualifiedName: "TestArray",
    });

    expect(printSourceData.props).toMatchSnapshot("array");

    expect(cache.size).toEqual(1);
  });
});
