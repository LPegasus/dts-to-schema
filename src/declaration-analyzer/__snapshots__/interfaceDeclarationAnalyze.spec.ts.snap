// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interface declaration analyze array.d.ts: array 1`] = `
Object {
  "arr1": Object {
    "isArray": true,
    "latitude": 1,
    "nullable": false,
    "optional": false,
    "v": "string",
  },
  "arr2": Object {
    "isArray": true,
    "latitude": 2,
    "nullable": false,
    "optional": false,
    "v": "string",
  },
  "arr3": Object {
    "isArray": true,
    "latitude": 1,
    "nullable": false,
    "optional": false,
    "v": "number",
  },
  "arr4": Object {
    "isArray": true,
    "latitude": 2,
    "nullable": false,
    "optional": false,
    "v": "number",
  },
  "arr5": Object {
    "isArray": true,
    "latitude": 2,
    "nullable": false,
    "optional": false,
    "v": "number",
  },
}
`;

exports[`interface declaration analyze base interface: base 1`] = `
Object {
  "foo": Object {
    "isArray": false,
    "nullable": false,
    "optional": false,
    "v": "string",
  },
  "optionalField": Object {
    "isArray": false,
    "nullable": false,
    "optional": true,
    "v": "number",
  },
}
`;
